### ISSUE TITLE
BUG: linalg.svd output doesn't match LAPACK results from C++ or R

### ISSUE THREAD

**Issue by graphic-goose** [scipy/scipy#18980]
### Describe your issue.

Hello,

I've been converting some C++ code to Python lately and part of the C++ code uses the LAPACK DGESVD function to calculate singular value decomposition.

I've called the scipy.linalg.svd function with lapack_driver='gesvd' but the results don't match what I get using the same matrix in both C++ and R functions that use LAPACK for SVD. 

For comparison, the R and C++ SVD functions return the same values as each other, but the Python does not.

This only seems to be an issue in the scenario where there's lots more columns than rows in the input matrix, which from the LAPACK doco seems to indicate it goes down a different path (- PATH 1t (N much larger than M, JOBVT='N')) involving LQ Decomposition.

Code in R:

```
> m <-10
> n <- 1000
> mat <- matrix(1, nrow=m, ncol=n) * 19
> svd(mat)
```

Code in C++:

```
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

// dgeev_ is a symbol in the LAPACK library files
extern "C" {
    extern void dgesvd_(char* jobu, char* jobvt, int* m, int* n, double* A,
        int* lda, double* S, double* U, int* ldu, double* VT,
        int* ldvt, double* work, int* lwork, int* info);
}

int main() {
    int m = 10; // Number of rows of the matrix A
    int n = 1000; // Number of columns of the matrix A
    int size = n * m;

    double* a, vt, *work;

    a = (double*)malloc(m * n * sizeof(double));

    for (int i = 0; i < size; ++i) {
        a[i] = 19;
    }

    int lda = m; // Leading dimension of the array A

    std::vector<double> s(std::min(m, n)); // Vector to store the singular values

    int ldu = m; // Leading dimension of the array U
    std::vector<double> u(ldu * m); // Array to store the left singular vectors U

    int ldvt = 1; // Leading dimension of the array VT
    vt = NULL;

    int lwork = m * 5; // Optimal work array size query flag
   
    int info; // Output info flag

    work = (double*)malloc(lwork * sizeof(double));

    double work_query; // Variable to query optimal work array size

    char save[] = "S", nosave[] = "N";

    // Compute the SVD
    dgesvd_(save, nosave, &m, &n, a, &lda, s.data(), u.data(), &ldu, &vt, &ldvt, work, &lwork, &info);

    // Check for successful computation
    if (info > 0) {
        std::cerr << "SVD computation failed: " << info << std::endl;
        return 1;
    }

    int array_length;
    array_length = m * m;

    // for each row
    for (int i = 0; i < array_length; i++)
    {
        cout << u.data()[i] <<"\n";
    }

    free(a);

    return 0;
}
```

### Reproducing Code Example

```python
import numpy as np
import scipy.linalg as linalg

m = 10
n = 1000
mat = np.ones((m, n))
mat = mat * 19
u, s, v = linalg.svd(mat, full_matrices=False, lapack_driver='gesvd')
```

### Error message

```shell
No error messages reported.
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.24.1 sys.version_info(major=3, minor=11, micro=3, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
  lapack:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.33
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-tkj0pjht\overlay\Lib\site-packages\pythran
    version: 0.12.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-7zsnxupn\cp311-win_amd64\build\venv\Scripts\python.exe
  version: '3.11'
```

**Comment by matthew-brett** [comment]
Are you sure your C++ and R code are linking to the same BLAS / LAPACK implementation as the Python code - i.e. the 32-bit OpenBLAS binary you see in the configuration output?

**Comment by graphic-goose** [comment]
@matthew-brett the `c:/opt/` path doesn't even exist on my computer, which is interesting...

The R code is using what was installed when I installed R and RStudio earlier today to compare.
The C++ is using what was installed using vcpkg through Visual Studio.

I've tried Scipy as pip-installed through another Python environment, as well as through Conda install through Anaconda.

**Comment by ilayn** [comment]
Could you please post the difference also? The singular values should be similar for not-very-small-values and the vectors can be different. The easiest test to check is to form the product `U@S@V.T` in all cases and check if they form `A` sufficiently close. There is always difference between different LAPACK runs and that is pretty normal depending on which branch the algorithm falls through.

**Comment by graphic-goose** [comment]
@ilayn the C++ code doesn't return V, so I've been comparing the values of U and S.

I've attached a spreadsheet comparing the values returned from R, C++ and Python for U and S. You can see the values match exactly for R and C++, but are quite different in Python.

I get dimensional errors trying to use `U@S@V.T`  to reconstruct for a matrix A of shape (10, 1000) in Python. I get an array back for using it in with a symmetric matrix, but the result doesn't equal the original A.

Using the following reconstructs the original array `mat` for symmertrical A, but returns broadcast errors for A of shape (10, 1000).
```
m = 3
n = 3
mat = np.ones((m, n))
mat = mat * 19
u, s, v = linalg.svd(mat, full_matrices=True, lapack_driver='gesvd')

Sigma = np.zeros((mat.shape[0], mat.shape[1]))
Sigma[:mat.shape[1], :mat.shape[1]] = np.diag(s)

recon_a = u.dot(Sigma.dot(v))
```

[SVD_Diffs.xlsx](https://github.com/scipy/scipy/files/12194228/SVD_Diffs.xlsx)
![U](https://github.com/scipy/scipy/assets/140783905/a6cae7ef-0db8-41c9-b5f1-64053889847b)
![S](https://github.com/scipy/scipy/assets/140783905/e32573b2-c53c-452d-9b7a-38961e14551e)

**Comment by ilayn** [comment]
> @ilayn the C++ code doesn't return V, so I've been comparing the values of U and S.

You can make it to return by turning the second `nosave` to `save` right?

Here is what I am getting from a random 10, 100 run. 

```python
import numpy as np
import scipy.linalg as la
m, n = 10, 100
rng = np.random.default_rng(1234567890987654321)  # Setting seed for reproducible a run
A = rng.uniform(-10, 10, size=[m, n])
u, s, v = la.svd(A, full_matrices=True, lapack_driver='gesvd')
Sigma = np.hstack([np.diag(s), np.zeros([m, n-m])])
np.max(np.abs(u @ Sigma @ v - A))
```

and this gives me `2.4868995751603507e-14` well within the bounds of numerical errors expected. 

So let's use your matrix of 19s; 

```python
import numpy as np
import scipy.linalg as la
m, n = 10, 100
A = 19*np.ones([m, n])
u, s, v = la.svd(A, full_matrices=True, lapack_driver='gesvd')
Sigma = np.hstack([np.diag(s), np.zeros([m, n-m])])
np.max(np.abs(u @ Sigma @ v - A))
```

This gives `5.3290705182007514e-14` again not much of a change in terms of errors. Repeating with `n = 1000` gives the error `1.787014980436652e-12` and that is acceptable for the problem size increase. 

If we check the singular values 

```python
1900.0000000000061
2.461171494043386e-12
7.379638718399482e-27
9.851269327816697e-41
3.049945205559884e-55
6.340240774708437e-70
2.5155503334025912e-84
5.743904953594653e-99
3.4129137230234146e-113
8.306586617239412e-128
```

So I don't see any mistake. However it might be that C++ and R are using different versions of LAPACK and that is quite ok to have different results. I'd suggest forming the products in the other code too and check the error bounds.

**Comment by graphic-goose** [comment]
I can confirm I get the same error bound as you in Scipy, of `2.4868995751603507e-14` when using `n = 100`. Doing the same in R given an error bound of `2.1316282072803006e-14`. 

In both scenarios calculating the product seems to faithfully reproduce the input array of 19's.

I had a go at tweaking the C++ code to output V to calculate the product, but it doesn't seem to be as simple as just changing the second `nosave` to `save`, as then it starts throwing an error about an illegal value for the `lwork` input to LAPACK. I tried different calculated values, from what I could tell from the doco, but still doesn't seem happy.

Interestingly enough, I found a way to call R through Python using the `rpy2` library, and while the values the SVD gives me matches the C++ result for my test matrix of 19's, once I throw the real data from the actual code I'm converting from C++ at it, the U values start to differ after the first column or two.

Either way, reading a bit more about SVD and working through your suggestions gives me more confidence that the SVD values I'm getting back are ok and it's just variances in versions of LAPACK.

Thanks for looking into it and providing the additional info.
